gobblin-core/src/main/java/gobblin/source/extractor/extract/QueryBasedSource.java

```diff
@@ -12,236 +12,245 @@
 import gobblin.source.extractor.JobCommitPolicy;
+import gobblin.source.extractor.WatermarkInterval;
 import gobblin.source.extractor.partition.Partitioner;
 import gobblin.source.extractor.utils.Utils;
-import gobblin.source.extractor.watermark.WatermarkPredicate;
-import gobblin.source.extractor.watermark.WatermarkType;
 import gobblin.source.workunit.Extract;
 import gobblin.source.workunit.Extract.TableType;
+import gobblin.source.workunit.MultiWorkUnit;
 import gobblin.source.workunit.WorkUnit;
+import gobblin.util.ConfigUtils;
+import gobblin.util.PathUtils;
+import gobblin.util.dataset.DatasetUtils;
+import lombok.extern.slf4j.Slf4j;
 
 
 /**
  * A base implementation of {@link gobblin.source.Source} for
  * query-based sources.
  */
+@Slf4j
 public abstract class QueryBasedSource<S, D> extends AbstractSource<S, D> {
 
-  private static final Logger LOG = LoggerFactory.getLogger(QueryBasedSource.class);
+  private static final String QUERY_BASED_SOURCE = "query_based_source";
 
   @Override
   public List<WorkUnit> getWorkunits(SourceState state) {
     initLogger(state);
 
     List<WorkUnit> workUnits = Lists.newArrayList();
     String nameSpaceName = state.getProp(ConfigurationKeys.EXTRACT_NAMESPACE_NAME_KEY);
-    String entityName = state.getProp(ConfigurationKeys.SOURCE_ENTITY);
 
-    String extractTableName = state.getProp(ConfigurationKeys.EXTRACT_TABLE_NAME_KEY);
-    // If extract table name is not found then use the entity name
-    if (StringUtils.isBlank(extractTableName)) {
-      extractTableName = Utils.escapeSpecialCharacters(entityName, ConfigurationKeys.ESCAPE_CHARS_IN_TABLE_NAME, "_");
+    Map<String, String> tableNameToEntityMap = Maps.newHashMap();
+    for (String entity : getSourceEntities(state)) {
+      tableNameToEntityMap.put(Utils.escapeSpecialCharacters(entity, ConfigurationKeys.ESCAPE_CHARS_IN_TABLE_NAME, "_"),
+          entity);
     }
 
-    TableType tableType = TableType.valueOf(state.getProp(ConfigurationKeys.EXTRACT_TABLE_TYPE_KEY).toUpperCase());
-    long previousWatermark = this.getLatestWatermarkFromMetadata(state);
-
-    Map<Long, Long> sortedPartitions = Maps.newTreeMap();
-    sortedPartitions.putAll(new Partitioner(state).getPartitions(previousWatermark));
+    Map<String, State> tableSpecificPropsMap = shouldObtainTablePropsFromConfigStore(state)
+        ? getTableSpecificPropsFromConfigStore(tableNameToEntityMap.keySet(), state)
+        : DatasetUtils.getDatasetSpecificProps(tableNameToEntityMap.keySet(), state);
+    Map<String, Long> prevWatermarksByTable = getPreviousWatermarksForAllTables(state);
 
-    // Use extract table name to create extract
-    SourceState partitionState = new SourceState();
-    partitionState.addAll(state);
-    Extract extract = partitionState.createExtract(tableType, nameSpaceName, extractTableName);
+    for (String tableName : Sets.union(tableNameToEntityMap.keySet(), prevWatermarksByTable.keySet())) {
```

Comment:
nit: iterating over the key set of a map is inefficient and requires a new set to be allocated. This will also trigger a FindBugs alerts. It is more efficient to iterate over the entry set.